### DISCUSSION TITLE
'NoneType' object has no attribute 'nominalWidthX' for empty glyph in FontBuilder

### DISCUSSION THREAD

**Discussion by NightFurySL2001** [fonttools/fonttools#3231] [General]
Is it normal for FontBuilder to throw a `'NoneType' object has no attribute 'nominalWidthX'` error when calculating bounds of an empty glyph? 
(E.g. passing an empty SVG to SVGPen and then use the charstring generated from it.) 

The code in relation is this which is given in the FontBuilder example:
```py
# calculat lsb
lsb = {gn: cs.calcBounds(None)[0] for gn, cs in glyphs.items()}
```
This raises `TypeError: 'NoneType' object is not subscriptable`, so when I try to isolate the issue it comes down to `cs.calcBounds(None)` gives `'NoneType' object has no attribute 'nominalWidthX'` for `cs` of empty glyph.

I'm building with OpenType/CFF. Slightly related to #1558 .

**Comment by TakWolf** [comment]
I also discovered the similar problem. On build `OTF`, invoke order seems important. That is to say, you must call 
```
        builder.setupCFF(cff_ps_name, cff_font_infos, glyphs, {})
```
first, and then can call 

```
    lsb = glyphs[glyph_name].calcBounds(None)[0]
```

Otherwise, it will cause

```
  File "/Users/takwolf/Develop/Python/pixel-font-builder/venv/lib/python3.11/site-packages/fontTools/misc/psCharStrings.py", line 1154, in calcBounds
    self.draw(boundsPen)
  File "/Users/takwolf/Develop/Python/pixel-font-builder/venv/lib/python3.11/site-packages/fontTools/misc/psCharStrings.py", line 1144, in draw
    self.private.nominalWidthX,
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'nominalWidthX'
```

Build `TTF` does not have this problem.

**Reply by NightFurySL2001** [reply]
Actually I'm aware of this ordering issue (#1558). The main problem is when the glyph is empty (like U+0020 space) the same error will happen even though the function call order is correct.

**Reply by TakWolf** [reply]
How did you create an empty glyph?  Like this?

```
        pen = T2CharStringPen(advance_width, None)
        pen.moveTo((0, 0))
        pen.closePath()
```

**Comment by NightFurySL2001** [comment]
Sorry forgot to attach a minimal reproducible code (modified from #1558). Here is the full code with a sample SVG file, edited lines are commented:
[empty.svg](https://github.com/fonttools/fonttools/assets/33471049/2873beef-ce08-4bde-b93d-9b7d9774e615)

```py
from fontTools.fontBuilder import FontBuilder
from fontTools.pens.t2CharStringPen import T2CharStringPen
from fontTools.svgLib import SVGPath


def drawTestGlyph(pen): # changed pen function to load svg
    svg_fulltext = open("empty.svg", "r", encoding="utf-8-sig").read()
    svgpen = SVGPath.fromstring(svg_fulltext)
    svgpen.draw(pen)

fb = FontBuilder(1024, isTTF=False)
fb.setupGlyphOrder([".notdef", ".null", "A", "a"])
fb.setupCharacterMap({65: "A", 97: "a"})

advanceWidths = {".notdef": 600, "A": 600, "a": 600, ".null": 600}

familyName = "HelloTestFont"
styleName = "TotallyNormal"
nameStrings = dict(familyName=dict(en="HelloTestFont", nl="HalloTestFont"),
                   styleName=dict(en="TotallyNormal", nl="TotaalNormaal"))
nameStrings['psName'] = familyName + "-" + styleName

pen = T2CharStringPen(600, None)
drawTestGlyph(pen)
charString = pen.getCharString()
charStrings = {".notdef": charString, "A": charString, "a": charString, ".null": charString}
fb.setupCFF(nameStrings['psName'], {"FullName": nameStrings['psName']}, charStrings, {})

metrics = {}
lsb = {gn: cs.calcBounds(None)[0] for gn, cs in charStrings.items()} # added line
for gn, advanceWidth in advanceWidths.items():
    metrics[gn] = (advanceWidth, lsb[gn])  # changed to use calculated lsb
fb.setupHorizontalMetrics(metrics)

fb.setupHorizontalHeader(ascent=824, descent=200)
fb.setupNameTable(nameStrings)
fb.setupOS2()
fb.setupPost()

fb.save("test.otf")
```
cc @TakWolf

**Reply by TakWolf** [reply]
This code cause `TypeError: 'NoneType' object is not subscriptable`

The empty svg actually did not provide any `program`(except [600, 'endchar']). Svg path has no executable transform.

This is equivalent to:

```
pen = T2CharStringPen(600, None)
pen.closePath()
```

I have the same problem.  I avoided this issue by adding a `pen.moveTo((0, 0))`

```
pen = T2CharStringPen(advance_width, None)
pen.moveTo((0, 0))
pen.closePath()
```

Now it seems that this is a bug and should be fixed.

**Reply by NightFurySL2001** [reply]
If you do `print(charStrings[0].calcBounds(None))` right before the `lsb` line I modified (tagged with comment), the error will be `'NoneType' object has no attribute 'nominalWidthX'`.

I believe this is a bug since in the CFF [Type 2 Charstring Format spec](https://adobe-type-tools.github.io/font-tech-notes/pdfs/5177.Type2.pdf) it stats that one width operator following by `endchar` is a valid charstring.

> _Note 7
> A character that does not have a path (e.g. a space character) may
> consist of an **endchar** operator preceded only by a width value.
> Although the width must be specified in the font, it may be specified as
> the **defaultWidthX** in the CFF data, in which case it should not be
> specified in the charstring. Also, it may appear in the charstring as the
> difference from **nominalWidthX**. Thus the smallest legal charstring
> consists of a single **endchar** operator._
